bee-crypto/src/ternary/sponge/batched_curlp/bct_curlp.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright 2020 IOTA Stiftung
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and limitations under the License.
+
+use crate::ternary::{
+    sponge::{
+        batched_curlp::{bct::BCTritBuf, HIGH_BITS},
+        CurlPRounds,
+    },
+    HASH_LENGTH,
+};
+
+pub struct BCTCurlP {
+    rounds: CurlPRounds,
+    state: BCTritBuf,
+    scratch_pad: BCTritBuf,
+}
+
+impl BCTCurlP {
+    #[allow(clippy::assertions_on_constants)]
+    pub fn new(rounds: CurlPRounds) -> Self {
+        // Ensure that changing the hash length will not cause undefined behaviour.
+        assert!(3 * HASH_LENGTH > 728);
```

Comment:
Should we return an `Err` if it was changed and no more valid?

Reply:
I think it should. `CurlP` has the error type of `Infallible` which might not be the same case in `BCTCurlP`.

Reply:
I don't think it should. This is not a recoverable issue, this is an implementation issue, there is nothing you can do at runtime if this is failing.

Reply:
It is also resolved during compilation, so basically this assertion does not appear in the binary if it is false. Otherwise the whole function gets compiled to a panic.